res/controllers/Behringer-CMDStudio4a-scripts.js

```diff
@@ -0,0 +1,257 @@
+    } else {
+        // No DEL so just cue.
+        engine.setValue(group, "cue_default", 1);
+    }
+}
+BehringerCMDStudio4a.cueRelease = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // DEL is active, reverse-roll release!
+        engine.setValue(group, "reverse", 0);
+    } else {
+        engine.setValue(group, "cue_default", 0);
+	}
+}
+
+
+// Function to deal with the scratch mode buttons.
+BehringerCMDStudio4a.scratch = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.scratchState[channel]) {
+        BehringerCMDStudio4a.scratchState[channel] = false;
+        midi.sendShortMsg(status, control, 0x00);
+    } else {
+        BehringerCMDStudio4a.scratchState[channel] = true;
+        midi.sendShortMsg(status, control, 0x01);
+    }
+}
+
+// Functions to deal with the hot-que buttons.
+BehringerCMDStudio4a.hotcuePush = function (channel, control, value, status, group) {
+    // Translate the button to the actual hotcue.
+    var hotcue = control-0x21;  // Hotcue buttons on left ddeck go from 0x22 to 0x29
+    if (hotcue>8) {
+        // Right deck, buttons are 0x20 higher so we need to compensate.
+        hotcue = hotcue-0x20;
+    }
+    print(hotcue)
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // Del "shift" is active, delete hotcue instead.
+        engine.setValue(group, "hotcue_"+hotcue+"_clear", 1);
+    } else {
+        // Del "shift" is not active, do the push hotcue function as normal.
+        engine.setValue(group, "hotcue_"+hotcue+"_activate", 1);
+    }
+}
+BehringerCMDStudio4a.hotcueRelease = function (channel, control, value, status, group) {
+    // Translate the button to the actual hotcue.
+    var hotcue = control-0x21;  // Hotcue buttons on left ddeck go from 0x22 to 0x29
+    if (hotcue>8) {
+        // Right deck, buttons are 0x20 higher so we need to compensate.
+        hotcue = hotcue-0x20;
+    }
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // Del "shift" is active, do nothing on button release.
+    } else {
+        // Del "shift" is not active, do the release hotcue function as normal.
+        engine.setValue(group, "hotcue_"+hotcue+"_activate", 0);
+    }
+}
+
+// Functions to deal with the pitch inc/dec buttons.
+BehringerCMDStudio4a.pitchDecPush = function (channel, control, value, status, group) {
+    BehringerCMDStudio4a.pitchDec[channel] = true;
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // Del "shift" is active, do "key" changes instead.
+        // Check if the other button is pressed too, if so we reset the pitch.
+        if (BehringerCMDStudio4a.pitchInc[channel]) {
+            engine.setValue(group, "pitch", 0);
+        } else {
+            engine.setValue(group, "pitch_down", 1);
+        }
+    } else {
+        // DEL mode is not active, just do "rate" changes.
+        engine.setValue(group, "rate_perm_down", 1);

```

Comment:
Use the rate_temp_down/up controls throughout the script. There's no point in having buttons for rate_perm_down/up when there is a pitch fader.